Data race in orderedNodeStream

I added make target `testrace`, which discovered the race below. I did a quick study of this, and found a few issues that I think is related to this. First, the following line is not protected by a `s.streamMut.RLock()`:

```go
	if s.streamBroken {
```

Similarly, lines like the one below should be protected by a write lock, not a read lock:

```go
	s.streamBroken = true
```

As such, it may or may not be worth it to use RLock for these issues. Maybe a plain mutex lock is fine?

Is it safe to protect these with a RLock only, or should they use a Lock?:

```go
	err := s.gorumsStream.SendMsg(req)
...
	err := s.gorumsStream.RecvMsg(resp)
```

```text
==================
WARNING: DATA RACE
Read at 0x00c0001543f8 by goroutine 16:
  github.com/relab/gorums/tests/reconnect.(*orderedNodeStream).sendMsgs()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:1008 +0x197

Previous write at 0x00c0001543f8 by goroutine 17:
  github.com/relab/gorums/tests/reconnect.(*orderedNodeStream).reconnectStream()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:1062 +0x463
  github.com/relab/gorums/tests/reconnect.(*orderedNodeStream).recvMsgs()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:1038 +0x24c

Goroutine 16 (running) created at:
  github.com/relab/gorums/tests/reconnect.(*orderedNodeStream).connectOrderedStream()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:994 +0x1f6
  github.com/relab/gorums/tests/reconnect.(*Node).connect()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:632 +0x458
  github.com/relab/gorums/tests/reconnect.(*Manager).connectAll()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:416 +0x1b7
  github.com/relab/gorums/tests/reconnect.NewManager()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:362 +0xf1a
  github.com/relab/gorums/tests/reconnect.TestReconnect()
      /Users/meling/work/gorums/tests/reconnect/reconnect_test.go:28 +0x24a
  testing.tRunner()
      /usr/local/Cellar/go/1.14.4/libexec/src/testing/testing.go:991 +0x1eb

Goroutine 17 (running) created at:
  github.com/relab/gorums/tests/reconnect.(*orderedNodeStream).connectOrderedStream()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:995 +0x235
  github.com/relab/gorums/tests/reconnect.(*Node).connect()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:632 +0x458
  github.com/relab/gorums/tests/reconnect.(*Manager).connectAll()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:416 +0x1b7
  github.com/relab/gorums/tests/reconnect.NewManager()
      /Users/meling/work/gorums/tests/reconnect/reconnect_gorums.pb.go:362 +0xf1a
  github.com/relab/gorums/tests/reconnect.TestReconnect()
      /Users/meling/work/gorums/tests/reconnect/reconnect_test.go:28 +0x24a
  testing.tRunner()
      /usr/local/Cellar/go/1.14.4/libexec/src/testing/testing.go:991 +0x1eb
==================
--- FAIL: TestReconnect (0.11s)
    testing.go:906: race detected during execution of test
FAIL
FAIL	github.com/relab/gorums/tests/reconnect	0.714s
```

--- a/cmd/protoc-gen-gorums/dev/ordering.go
+++ b/cmd/protoc-gen-gorums/dev/ordering.go
@@ -73,7 +73,7 @@ type orderedNodeStream struct {
 	gorumsClient ordering.GorumsClient
 	gorumsStream ordering.Gorums_NodeStreamClient
 	streamMut    sync.RWMutex
-	streamBroken bool
+	streamBroken uint32
 }
 
 func (s *orderedNodeStream) connectOrderedStream(ctx context.Context, conn *grpc.ClientConn) error {
@@ -97,7 +97,7 @@ func (s *orderedNodeStream) sendMsgs(ctx context.Context) {
 		case req = <-s.sendQ:
 		}
 		// return error if stream is broken
-		if s.streamBroken {
+		if atomic.LoadUint32(&s.streamBroken) == 1 {
 			err := status.Errorf(codes.Unavailable, "stream is down")
 			s.putResult(req.metadata.MessageID, &orderingResult{nid: s.node.ID(), reply: nil, err: err})
 			continue
@@ -109,7 +109,7 @@ func (s *orderedNodeStream) sendMsgs(ctx context.Context) {
 			s.streamMut.RUnlock()
 			continue
 		}
-		s.streamBroken = true
+		atomic.StoreUint32(&s.streamBroken, 1)
 		s.streamMut.RUnlock()
 		s.node.setLastErr(err)
 		// return the error
@@ -123,7 +123,7 @@ func (s *orderedNodeStream) recvMsgs(ctx context.Context) {
 		s.streamMut.RLock()
 		err := s.gorumsStream.RecvMsg(resp)
 		if err != nil {
-			s.streamBroken = true
+			atomic.StoreUint32(&s.streamBroken, 1)
 			s.streamMut.RUnlock()
 			s.node.setLastErr(err)
 			// attempt to reconnect
@@ -151,7 +151,7 @@ func (s *orderedNodeStream) reconnectStream(ctx context.Context) {
 		var err error
 		s.gorumsStream, err = s.gorumsClient.NodeStream(ctx)
 		if err == nil {
-			s.streamBroken = false
+			atomic.StoreUint32(&s.streamBroken, 0)
 			return
 		}
 		s.node.setLastErr(err)

--- a/cmd/protoc-gen-gorums/gengorums/template_static.go
+++ b/cmd/protoc-gen-gorums/gengorums/template_static.go
@@ -1007,7 +1007,7 @@ type orderedNodeStream struct {
 	gorumsClient ordering.GorumsClient
 	gorumsStream ordering.Gorums_NodeStreamClient
 	streamMut    sync.RWMutex
-	streamBroken bool
+	streamBroken uint32
 }
 
 func (s *orderedNodeStream) connectOrderedStream(ctx context.Context, conn *grpc.ClientConn) error {
@@ -1031,7 +1031,7 @@ func (s *orderedNodeStream) sendMsgs(ctx context.Context) {
 		case req = <-s.sendQ:
 		}
 		// return error if stream is broken
-		if s.streamBroken {
+		if atomic.LoadUint32(&s.streamBroken) == 1 {
 			err := status.Errorf(codes.Unavailable, "stream is down")
 			s.putResult(req.metadata.MessageID, &orderingResult{nid: s.node.ID(), reply: nil, err: err})
 			continue
@@ -1043,7 +1043,7 @@ func (s *orderedNodeStream) sendMsgs(ctx context.Context) {
 			s.streamMut.RUnlock()
 			continue
 		}
-		s.streamBroken = true
+		atomic.StoreUint32(&s.streamBroken, 1)
 		s.streamMut.RUnlock()
 		s.node.setLastErr(err)
 		// return the error
@@ -1057,7 +1057,7 @@ func (s *orderedNodeStream) recvMsgs(ctx context.Context) {
 		s.streamMut.RLock()
 		err := s.gorumsStream.RecvMsg(resp)
 		if err != nil {
-			s.streamBroken = true
+			atomic.StoreUint32(&s.streamBroken, 1)
 			s.streamMut.RUnlock()
 			s.node.setLastErr(err)
 			// attempt to reconnect
@@ -1085,7 +1085,7 @@ func (s *orderedNodeStream) reconnectStream(ctx context.Context) {
 		var err error
 		s.gorumsStream, err = s.gorumsClient.NodeStream(ctx)
 		if err == nil {
-			s.streamBroken = false
+			atomic.StoreUint32(&s.streamBroken, 0)
 			return
 		}
 		s.node.setLastErr(err)

